gobblin-data-management/src/main/java/org/apache/gobblin/data/management/copy/iceberg/IcebergTableMetadataValidatorUtils.java

```diff
@@ -0,0 +1,96 @@
+
+import lombok.extern.slf4j.Slf4j;
+
+/**
+ * Validator for Iceberg table metadata, ensuring that the source and destination tables have
+ * compatible schemas and partition specifications.
+ */
+@Slf4j
+public class IcebergTableMetadataValidatorUtils {
+
+  private IcebergTableMetadataValidatorUtils() {
+    // Do not instantiate
+  }
+
+  /**
+   * Validates the metadata of the source and destination Iceberg tables.
+   *
+   * @param srcTableMetadata  the metadata of the source table
+   * @param destTableMetadata the metadata of the destination table
+   * @throws IOException if the schemas or partition specifications do not match
+   */
+  public static void validateSourceAndDestinationTablesMetadata(TableMetadata srcTableMetadata,
+      TableMetadata destTableMetadata) throws IOException {
+    log.info("Starting validation of Source : {} and Destination : {} Iceberg Tables Metadata",
+        srcTableMetadata.location(),
+        destTableMetadata.location());
+    Schema srcTableSchema = srcTableMetadata.schema();
+    Schema destTableSchema = destTableMetadata.schema();
+    PartitionSpec srcPartitionSpec = srcTableMetadata.spec();
+    PartitionSpec destPartitionSpec = destTableMetadata.spec();
+    validateSchemaForEquality(srcTableSchema, destTableSchema);
+    validatePartitionSpecForEquality(srcPartitionSpec, destPartitionSpec);
+    log.info("Validation of Source : {} and Destination : {} Iceberg Tables Metadata completed successfully",
+        srcTableMetadata.location(),
+        destTableMetadata.location());
+  }
+
+  private static void validateSchemaForEquality(Schema srcTableSchema, Schema destTableSchema) throws IOException {
+    // TODO: Need to add support for schema evolution, currently only supporting copying
+    //  between iceberg tables with same schema.
+    //  This function needs to be broken down into multiple functions to support schema evolution
+    //  Possible cases - Src Schema == Dest Schema,
+    //  - Src Schema is subset of Dest Schema [ Destination Schema Evolved ],
+    //  - Src Schema is superset of Dest Schema [ Source Schema Evolved ],
+    //  - Other cases?
+    //  Also consider using Strategy or any other design pattern for this to make it a better solution
+    if (!srcTableSchema.sameSchema(destTableSchema)) {
+      String errMsg = String.format(
+          "Schema Mismatch between Source and Destination Iceberg Tables Schema - Source-Schema-Id : {%s} and "
+              + "Destination-Schema-Id : {%s}",
+          srcTableSchema.schemaId(),
+          destTableSchema.schemaId()
+      );
+      log.error(errMsg);
+      throw new IOException(errMsg);
+    }
+  }
+
+  private static void validatePartitionSpecForEquality(PartitionSpec srcPartitionSpec, PartitionSpec destPartitionSpec)
+      throws IOException {
+    // Currently, only supporting copying between iceberg tables with same partition spec
+    if (!srcPartitionSpec.compatibleWith(destPartitionSpec)) {
```

Comment:
the method name says equality.  is that the same as `compatibleWith`?  it sounds more lenient than strict equality.

Reply:
The difference b/w `compatibleWith` & `equals` method in PartitionSpec is that  `compatibleWith` doesn't match for specId while `equals` does, given that both tables are different matching equality for specId doesn't seem to be correct. 

Please let me know your thoughts around this.

Reply:
merged all the method into single method `failUnlessCompatibleStructure(TableMetadata tableAMetadata, TableMetadata tableBMetadata)` as having those separate methods just for one check doesn't seem to be a correct approach and also after merging able to print metadatafilelocation with both the IOException.

Reply:
are you saying we want to verify names match in addition to structure?  if so, I believe that's what's required for reading the underlying orc files... but I've never verified.